File: 713-subarray-product-less-than-k/subarray-product-less-than-k.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int count=0;
        
        long long product=1;
        if(k<=1){
            return 0;
        }
        // two pointer
        int start=0;
        int end =0;
        while(end<nums.size()){
            product *=nums[end];

            while(product>=k){
                product=product/nums[start];
                start++;
            }

            count = count + end-start + 1;
            end++; 
        }
        return count;
    }
};

File: 713-subarray-product-less-than-k/subarray-product-less-than-k.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        int count = 0;
        int prod = 1;
        int i= 0;
        for(int j=0;j<nums.length;j++){
            prod = prod*nums[j];
            while(prod>=k && i <= j){
                prod = prod/nums[i];
                i++;
            }
            count += j-i+1;
        }
        return count;
    }
}